File: fl/implementations/haskell/Evaluator.hs
module Evaluator where

import Prelude hiding (lookup)
import Data.Map
import Types

evaluate :: [Function] -> Expr -> Maybe Int
evaluate fs e = evaluate empty e
  where
    evaluate m (Literal i) = Just i
    evaluate m (Argument n) = lookup n m
    evaluate m (Binary o x y) = do 
    	     x' <- evaluate m x
	     y' <- evaluate m y
             return (case o of
               Equal -> if x'==y' then -1 else 0
               Plus -> x' + y'
               Minus -> x' - y') 
    evaluate m (IfThenElse x y z) = do 
    	     x' <- evaluate m x
	     if x' /= 0 then evaluate m y else evaluate m z
    evaluate m (Apply n es) = do
    	     is <- mapM (evaluate m) es
    	     (ns,e) <- head ([ Just (ns,e) | (Function (n',ns) e) <- fs, n == n' ] ++ [Nothing])
	     let m' = fromList (zip ns is)
	     evaluate m' e

File: fl/implementations/java/Evaluator.java
import java.util.*;

/**
 * A visitor for evaluation
 */
public class Evaluator extends Visitor {
	private int result;
	private LinkedList<Function> functions;
	private Stack<HashMap<String,Integer>> stack;
	public Evaluator(Program program) {
	    functions = program.functions;
	    stack = new Stack<HashMap<String,Integer>>();
	}
	public int evaluate(Expr e) {
	    e.accept(this);
	    return result;
	}
	public void visit(Program x) {
	    throw new EvaluatorException();
	}
	public void visit(Function x) {
	    throw new EvaluatorException();
	}
	public void visit(Literal x) {
	    result = x.info;
	}
	public void visit(Argument x) {
	    result = stack.peek().get(x.name);
	}
	public void visit(Binary x) {
	    int a = evaluate(x.left);
	    int b = evaluate(x.right);
	    switch (x.ops) {
	       case Equal : 
		   result = a == b ? -1 : 0;
		   break;
	       case Plus : 
		   result = a + b;
		   break;
	       case Minus : 
		   result = a - b;
		   break;
	    }
	}
	public void visit(IfThenElse x) {
	    result = evaluate(x.exprIf) != 0 ?
		  evaluate(x.exprThen)
		: evaluate(x.exprElse);
	}
	public void visit(Apply x) {
	    Function f = null;
	    for (Function g : functions)
		if (g.name.equals(x.name)) {
		    f = g;
		    break;
		}
	    HashMap<String,Integer> map = new HashMap<String,Integer>();
	    Iterator<String> i = f.args.iterator();
	    for (Expr e : x.args)
		map.put(i.next(),evaluate(e));
	    stack.push(map);
	    f.rhs.accept(this);
	    stack.pop();
	}
}
